File: 2023/ts/06/part01.ts
import fs from "fs";
import path from "path";
import { range, zip } from "../utils";

{
  //   const text = `
  // Time:      7  15   30
  // Distance:  9  40  200
  // `;

  const text = fs.readFileSync(path.join(__dirname, "./input.txt"), {
    encoding: "ascii",
  });

  const lines = text.trim().split("\n");
  const times = lines[0]!.split(/:\s+/)[1]!.split(/\s+/).map(Number);
  const distances = lines[1]!.split(/:\s+/)[1]!.split(/\s+/).map(Number);

  const sum = zip(times, distances)
    .map(([t, d]) => {
      const cnt = range(1, t)
        .map((holding) => {
          const moving = t - holding;
          const dist = moving * holding;
          return Number(dist > d);
        })
        .reduce((acc, v) => acc + v, 0);
      return cnt;
    })
    .reduce((acc, v) => acc * v, 1);

  console.log(sum); // 345015
}

File: 2023/rust/day06/src/part01.rs
use regex::Regex;
use std::fs;

pub fn solve() {
    //     let text = "Time:      7  15   30
    // Distance:  9  40  200
    // ";

    let text = fs::read_to_string("input.txt").unwrap();

    let lines = text.trim().split("\n").collect::<Vec<_>>();
    let times = Regex::new(r"\s+")
        .unwrap()
        .split(lines[0])
        .collect::<Vec<_>>()[1..]
        .into_iter()
        .map(|s| s.parse::<i32>().unwrap())
        .collect::<Vec<_>>();
    let dists = Regex::new(r"\s+")
        .unwrap()
        .split(lines[1])
        .collect::<Vec<_>>()[1..]
        .into_iter()
        .map(|s| s.parse::<i32>().unwrap())
        .collect::<Vec<_>>();

    let res = times
        .into_iter()
        .zip(dists)
        .map(|(t, d)| {
            (1..t)
                .map(|holding| {
                    let moving = t - holding;
                    let dist = moving * holding;
                    (dist > d) as i32
                })
                .sum::<i32>()
        })
        .product::<i32>();

    println!("{res}"); // 345015
}
